packages/data/src/components/use-select/index.js

```diff
@@ -227,7 +230,10 @@ export default function useSelect( _mapSelect, deps ) {
 			unsubscribers.forEach( ( unsubscribe ) => unsubscribe?.() );
 			renderQueue.flush( queueContext );
 		};
-	}, [ registry, trapSelect, depsChangedFlag, isWithoutMapping ] );
+		// If you're tempted to eliminate the spread dependencies below don't do it!
+		// We're passing these in from the calling function and want to make sure we're
+		// examining every individual value inside the `deps` array.
+	}, [ registry, trapSelect, hasMappingFunction, ...( deps ?? [] ) ] );
```

Comment:
I have some doubts whether we really want `deps` as a dependency of this effect. The effect's purpose is to subscribe to the stores that the `mapSelect` function uses (are in the `listeningStores` list) and resubscribe if the `listeningStores` changes.

What if we use `useSelect` like this?
```js
useSelect( ( select ) => {
  return select( 'a' ).get() || select( 'b' ).get();
}, [] );
```
Then, if on the first call the hook selects only from `a` and doesn't select from `b`, we'll subscribe only to `a`. And these subscriptions will never change during the lifetime of the hook, because the `deps` are empty and there is nothing that would trigger the re-subscription effect. Now, when `select( 'a' ).get()` updates and starts returning `false`, we'll select from `b` only once and won't react to changes in `b`, because we're not subscribed to it.

The re-subscription hook should somehow depend on the `listeningStores` array instead.

If I write the hook without dependencies:
```js
useSelect( ( select ) => {
  return select( 'a' ).get() || select( 'b' ).get();
} );
```
the situation is exactly the same, the only difference is that the `mapSelect` function won't get memoized and will be called more often.

Reply:
I agree that having `deps` in this hook might be a mistake. Redux's [`useSelector`](https://react-redux.js.org/api/hooks#useselector) doesn't have that either. The gotcha here is that we need to make sure the `mapSelect` function is a pure function (without side-effects), but I think we already only support that.

We can do some performance audit to see if removing the deps has any impacts. But not in this PR I guess 😅 .

Reply:
> we need to make sure the `mapSelect` function is a pure function

The `mapSelect` function in not generally pure, there are perfectly valid cases like this one mentioned in README:
```js
function HammerPriceDisplay( { currency } ) {
  const price = useSelect( ( select ) => {
    return select( 'my-shop' ).getPrice( 'hammer', currency );
  }, [ currency ] );
  /* ... */
}
```
That's the canonical use case for the deps: we need to rerun the `mapSelect` function not only when store state changes, but also when the `currency` dep changes. The result depends on both.

> Redux's `useSelector` doesn't have that either.

That's true, our `useSelect` does some optimizations, like avoiding the call to `mapSelect` when the last result can be reused, while Redux `useSelector` doesn't care about that. In
```js
useSelector( ( state ) => getPrice( state, 'hammer', currency ) );
```
the selector function has a new instance on every render, and is called on every render.

Reply:
> That's the canonical use case for the deps: we need to rerun the mapSelect function not only when store state changes, but also when the currency dep changes. The result depends on both.

Isn't this also a pure function though? By side-effects-free, I mean the results will be the same for the given inputs no matter how many times we run it. The same as how React's render function is pure. An example of making it not pure would be like adding a `console.log` in it.

Reply:
I see, "pure" is probably not a good term to describe what's going on 🙂 The point is that without the `currency` dependency:
```js
const price = useSelect( ( select ) => {
  return select( 'my-shop' ).getPrice( 'hammer', currency );
}, [] );
```
the `useSelect` hook would store the `mapSelect` function from the first render to a ref, and the functions passed on subsequent renders would be ignored, although they are different. And they may return different results because `currency` has changed.

Reply:
> I mean the results will be the same for the given inputs no matter how many times we run it.

That's not actually the case, as @jsnajdr pointed out. It's still correct to say then that it's not pure because its output does depend on globals that aren't specified as inputs, apart from the fact that we sneak it in through `select`. So arguably sure, `select` is technically an input, which itself could encapsulate the state of the world when called, but that's not really helpful here where we're asking more pointedly, "can we run this numerous times without messing things up?" and "can we safely return a cached version of this output?"

which brings us to this…

> The re-subscription hook should somehow depend on the listeningStores array instead.

totally agree, but do'm not sure that's possible either because of how we've written the function, even with `__experimentalMarkListeningStores`.

```js
const price = useSelect( ( select ) => {
    const couponCode = select( 'my-shop' ).currentPromotion();
    return couponCode !== null
        ? select( 'coupon-store' ).getPrice( couponCode, 'hammer' )
        : select( 'my-shop' ).getPrice( 'hammer' );
}, [] );
```

that's silly of course, but like the rule-of-hooks we depend here on people manually doing things right and never conditionally-calling a store if we want to make that kind of assertion.

are there known problems with the current behavior? is it the case that if people realize they aren't getting proper updates that they simply find and add the dependency to the `useSelect( …, deps )` call?

if that's the case, seems like `deps` works fine.

> I think that depsChangedFlag is a smart and correct workaround for this and I would personally keep it.

can you share what about this feels correct to you? I considered it a workaround because of the way we're creating a new thing when we could have spread the original to the same effect. whether or not those deps belong in there being another question. it's a surprising code pattern that works incidentally because `useMemo` is doing the same thing as `useLayoutEffect` but with `depsChangedFlag` the deps are "already spread" when passing `deps` to `useLayoutEffect` is _in effect_ wrapping the deps in a new array (which hides them). or at least this is the line of thinking which made me ask about removing it.

what I don't like as much about using `depsChangedFlag`, _especially_ if the reason is to avoid the lint warning, is that we can make the warning disappear but we'd actually still be preserving the behavior the lint was design to push against: nothing has changed except for added runtime overhead (`useMemo`) and now the warning is hidden (much like how we've hidden the warning for `useIsomorphicLayoutEffect` even though we're still presenting the same danger).

I don't have a strong opinion here or in this PR. I would lean on not changing things if we don't find agreement. it's just that I stumbled upon it and I think, based on the discussion, that my suspicion was correct - we created this workaround because without flattening or spreading the `deps` in `useEffect` it wasn't actually examining those dependencies, it was examining the array that held them, which was often needlessly new because of things like passing literal `[]`s which always creates a new array instance - and it seems like a good place to clean it up if it's producing the original intended effect - base the dependencies on these things _plus whatever the caller passed as `deps`_, which again, I think it does.

Reply:
> can you share what about this feels correct to you?

By _correct_ I mean that the code works correctly, i.e., `depsChangedFlag` changes if and only if any member of the `deps` array changes.

And it's a nice way how to compose dependency arrays, where:
```js
useEffect( ..., [ ...deps1, ...deps2 ] );
```
is equivalent to:
```js
const deps1changed = useMemo( {}, deps1 );
const deps2changed = useMemo( {}, deps2 );
useEffect( ..., [ deps1changed, deps2changed ] );
```
In the end, it's mostly a stylistic choice, similar to choosing between writing an expression inline and extracting it to a variable, and I don't have a strong opinion about it. Use the `...deps` spread if you like it more.

Reply:
Instead of passing `deps` as dependency, one way of another, we could simply pass the `callbackMapper`, because it only changes when `deps` change. And it conveys the intent better: we want to resubscribe when the `callbackMapper` changes, because that could lead to change in `listeningStores`.

And we also need to cover for the (legacy?) case where `useSelect` is used without `deps`:
```js
useSelect( () => {
  /* ... */
} );
```
because in that case we _never_ want to resubscribe, because the callback is different on each call and we'd be re-subscribing on each render. All written together:
```js
const mapperCallback = useCallback( ..., deps );
const resubscribeFlag = deps ? mapperCallback : null;
useEffect( () => {
  /* resubscibe */
}, [ ..., resubscribeFlag ] );
```
If the `useSelect` hook is called without `deps`, then the `resubscribeFlag` is always `null` and will never trigger re-subscription.

Reply:
> are there known problems with the current behavior? is it the case that if people realize they aren't getting proper updates that they simply find and add the dependency to the `useSelect( …, deps )` call?

I don't know if there are known problems with that, but it's quite plausible that we already have unnoticed bugs like this. The buggy behavior would be rather subtle -- a missed React state update here and there. Nothing the user can put their finger on and provide reliable steps to reproduce. Just the editor acting a bit erratically at times.

But it would be a good rule to document and check with a custom ESLint rule: in the `useSelect` callback, call `select` unconditionally, outside loops, and always in the same order. Just like the rule for hooks.

Reply:
We're already using `useCallback` to memoize the `mapSelect` function, which is not a [semantic guarantee](https://reactjs.org/docs/hooks-reference.html#usememo). I actually think it's a bug and an unexpected behavior to not always run the latest `mapSelect` function.

I still prefer to remove `deps` altogether, and if there's a case that they really want to memoize `mapSelect`, they can still do that by using `useCallback` themselves.

I don't think we have to enforce any other _rules_ besides ensuring the function is _pure_ either. `__experimentalMarkListeningStores` is designed to only subscribe to walked stores, that's intentional. Users shouldn't worry about any other implementation details. Selectors may or may not re-run several times during renders, but that's okay, as long as the selectors are _pure_.

Reply:
> I actually think it's a bug and an unexpected behavior to not always run the latest `mapSelect` function.

I understand it as an optimization. The `deps` tell `useSelect` that the `mapSelect` function can be memoized. I.e., when the `deps` haven't changed, and `state` hasn't changed either, then `mapSelect` won't be run: the hook will return the remembered value instead. For example:
```js
const [ tick, setTick ] = useState( 0 );
const price = useSelect( ( select ) => select( 'cart' ).getPrice(), [] );
```
If there were no `[]` deps array, the selector function would be called on each render, i.e., on every `setTick` call, although the data store hasn't changed. With deps, the `price` will be memoized.

Redux `useSelector` does a similar optimization, but less ambitious: only if the selector function is identical to the previous one, it will memoize. If the selector is defined inline (`useSelector( state => ... )`), it will be always called.

> if there's a case that they really want to memoize `mapSelect`, they can still do that by using `useCallback` themselves.

Yes, that's what you have to do with Redux `useSelector` if you want to do the nontrivial optimization:
```js
const selector = useCallback( ( state ) => getPrice( state, currency ), [ currency ] );
const price = useSelector( selector );
```

The only caveat is that we use the moment when `mapSelect` changes as a (heuristic) hint that we should resubscribe to stores, too. If someone writes:
```js
useSelect( ( select ) => select( 'cart' ).getPrice() )
```
they probably think "I don't care about microoptimizing the selector and I'm fine with it being called all the time", but what's bad is that the hook would also unsubscribe and resubscribe to the `cart` store on each call.

We'd need to compare the `listeningStores` sets on each call and modify subscriptions if they differ.

> We're already using useCallback to memoize the mapSelect function, which is not a semantic guarantee.

I think this is OK because if `useCallback`/`useMemo` forgets the memoized value then the component is merely deoptimized, executing more code than needed, but the behavior continues to be correct.

That's what the "semantic guarantee" warning means: would your component still work if `useMemo` started forgetting things? If it would break, you're using `useMemo` wrong.

Reply:
this discussion has led way deeper than I want to touch in this particular PR, and I appreciate that, but I want to cut the scope so we can attend to these problems in focused work.

for now I'm wanting to either leave things the way they were _or_ replace `useMemo` with `...deps` (which I think is the only change that preserves the existing behavior and semantics while eliminating needless runtime overhead).

any preference from you two on whether to go ahead and remove the overhead or leave it in?

Reply:
As I commented earlier, I like the `useMemo` "deps composition" trick and would personally use it. But `...deps` is fine, too. Going to approve this PR now, as my only blocking feedback (`useCallback` param type) has been addressed.

Reply:
thanks @jsnajdr - I went ahead and removed this change from the PR. it sounds like we all agree there's more behavioral work to be done on this one and I don't want to muddy this change with that; if we're thinking of touching it later we can do it in one go and preserve the commit history anyway.